### PR TITLE
Nats

### PR DESCRIPTION
**Pull request by Wectro20**

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -20,6 +20,10 @@ jobs:
 
       - uses: actions/checkout@v3
 
+      - name: Setup Protobuf
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y protobuf-compiler
```
Comment (ajax-gnatiuk-s) [inline .github/workflows/gradle.yml:26]:
```suggestion
```

We don't need this step. Ask @OlehSarapuk

Comment (OlehSarapuk) [reply .github/workflows/gradle.yml:26]: Your right
There are all needed dependencies and plugin to do it automatically

Comment (Wectro20) [reply .github/workflows/gradle.yml:26]: @ajax-gnatiuk-s 
Ur right, I'll rework this

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyPagesControllerTest.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                cryptocurrencyService.getCryptocurrencyPages(cryptoName, 1, 10)
+            }
+
+            sortedList.forEach { crypto ->
+                verify {
+                    cryptocurrencyConvertor.cryptocurrencyToProto(crypto)
+                }
+            }
+        }
+    }
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyPagesControllerTest.kt:98]:
```suggestion
    @ParameterizedTest
    @ValueSource(strings = ["crypto 1", "crypto 2", "crypto 3"]) // substitute your values 
    fun testHandler(cryptoName: String) {
            val sortedList = cryptocurrencyList.filter { it.cryptocurrencyName == cryptoName }

            val request = CryptocurrencyRequest.newBuilder()
                .setPage(
                    CryptocurrencyOuterClass.CryptocurrencyPage.newBuilder()
                        .setName(cryptoName)
                        .setPageNumber(1)
                        .setPageSize(10)
                )
                .build()

            every {
                cryptocurrencyService.getCryptocurrencyPages(cryptoName, 1, 10)
            } returns sortedList

            sortedList.forEach { crypto ->
                every {
                    cryptocurrencyConvertor.cryptocurrencyToProto(crypto)
                } returns crypto.toProto()
            }

            val response = controller.handler(request)
            val cryptoListFromResponse =
                response.cryptocurrencyList.cryptocurrencyList.map { it.toDomain() }

            assertEquals(sortedList.map { it.cryptocurrencyName }, cryptoListFromResponse.map { it.cryptocurrencyName })
            assertEquals(sortedList.map { it.price }, cryptoListFromResponse.map { it.price })
            assertEquals(sortedList.map { it.createdTime }, cryptoListFromResponse.map { it.createdTime })

            verify {
                cryptocurrencyService.getCryptocurrencyPages(cryptoName, 1, 10)
            }

            sortedList.forEach { crypto ->
                verify {
                    cryptocurrencyConvertor.cryptocurrencyToProto(crypto)
                }
            }
    }
```

Comment (ajax-gnatiuk-s) [reply src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyPagesControllerTest.kt:98]: please, replace all your for-each loops in tests with `ParameterizedTest`

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetCsvController.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    }
+
+//    val names = CryptocurrencyOuterClass.CryptocurrencyName.newBuilder().setName("cryptocurrency")
+//
+//    val a = CryptocurrencyRequest.newBuilder().setName(names).build()
+//    val b = natsConnection.requestWithTimeout("v2.input.reqreply.company.cryptocurrency.get_csv", a.toByteArray(), Duration.ofMillis(10000))
+//    val responseBytes = b.get().data
+//    val response = CryptocurrencyOuterClass.CryptocurrencyResponse.parseFrom(responseBytes)
+//    println(response)
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetCsvController.kt:74]: don't commit commented code

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        cryptocurrencyService = mockk()
+        cryptocurrencyConvertor = mockk()
+        connection = mockk()
+
+        controller = NatsCryptocurrencyGetAllController(
+            cryptocurrencyService,
+            cryptocurrencyConvertor,
+            connection
+        )
+    }
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt:55]:
```suggestion
    @MockK
    private lateinit var cryptocurrencyService: CryptocurrencyService
    
    @MockK
    private lateinit var cryptocurrencyConvertor: CryptocurrencyConvertor
    
    @MockK
    private lateinit var connection: Connection
    
    @InjectMockks
    private lateinit var controller: NatsCryptocurrencyGetAllController

    private val time = OffsetDateTime.now(ZoneOffset.UTC).toLocalDateTime()
    private val id: ObjectId = ObjectId("63b346f12b207611fc867ff3")

    private val cryptocurrencyList = listOf(
        Cryptocurrency(id, "BTC", 12341f, time),
        Cryptocurrency(id, "BTC", 23455f, time),
        Cryptocurrency(id, "ETH", 1200f, time),
        Cryptocurrency(id, "ETH", 1300f, time),
        Cryptocurrency(id, "ETH", 1400f, time),
        Cryptocurrency(id, "XRP", 200f, time),
        Cryptocurrency(id, "XRP", 300f, time),
        Cryptocurrency(id, "XRP", 520f, time)
    )
```

File: src/test/kotlin/com/ajax/cryptocurrency/controller/CryptocurrencyControllerTest.kt [hunk 5]
```diff
@@ -44,25 +44,25 @@ class CryptocurrencyControllerTest(@Value("\${cryptocurrency.name}") private val
...
+        Cryptocurrency(id, "XRP", 200f, time),
+        Cryptocurrency(id, "XRP", 300f, time),
+        Cryptocurrency(id, "XRP", 520f, time)
     )
 
     @Test
     fun getPriceTests() {
         for (crypto in cryptocurrencies) {
-            val cryptocurrencyPrice = Cryptocurrency(objectId, crypto, 12341f, time)
+            val cryptocurrencyPrice = Cryptocurrency(ObjectId("63b346f12b207611fc867ff3"), crypto, 12341f, time)
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/controller/CryptocurrencyControllerTest.kt:65]:
```suggestion
            val cryptocurrencyPrice = Cryptocurrency(id, crypto, 12341f, time)
```

?

File: nats-api/src/main/proto/com/ajax/cryptocurrency/cryptocurrency.proto [hunk 6]
```diff
@@ -0,0 +1,42 @@
+syntax = "proto3";
+
```
[the remark this answers is not in the document]
Comment (Wectro20) [reply nats-api/src/main/proto/com/ajax/cryptocurrency/cryptocurrency.proto:2]: done

File: src/main/kotlin/com/ajax/cryptocurrency/config/NatsListenerBeanPostProcessor.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+class NatsListenerBeanPostProcessor : BeanPostProcessor {
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any {
+        if (bean is NatsController<*, *>) {
+            bean.initializeNatsController()
+        }
+        return bean
+    }
+
+    private fun <ReqT : GeneratedMessageV3, RepT : GeneratedMessageV3> NatsController<ReqT, RepT>
+            .initializeNatsController() {
```
[the remark this answers is not in the document]
Comment (Wectro20) [reply src/main/kotlin/com/ajax/cryptocurrency/config/NatsListenerBeanPostProcessor.kt:18]: done

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -7,3 +7,7 @@ spring:
 cryptocurrency:
   name: BTC,ETH,XRP
   sleepTime: 5000
+
+nats:
+  connection:
+    url: nats://localhost:4222
```
Comment (OlehSarapuk) [inline src/main/resources/application.yml:13]: will it run with docker on my machine?

**Review by OlehSarapuk — COMMENTED** [review]
Well done!)

File: build.gradle.kts [hunk 9]
```diff
@@ -30,10 +32,14 @@ dependencies {
 	implementation("org.springframework.boot:spring-boot-starter-web")
 	implementation("org.jetbrains.kotlin:kotlin-reflect")
 	implementation("org.json:json:20230227")
+	implementation("io.nats:jnats:2.16.14")
+	implementation("com.google.protobuf:protobuf-java:3.24.2")
+	implementation("com.google.protobuf:protobuf-java-util:3.20.1")
+	implementation(project(":nats-api"))
+
 	annotationProcessor("org.springframework.boot:spring-boot-configuration-processor")
+	implementation("org.springframework.boot:spring-boot-starter-validation")
```
[the remark this answers is not in the document]
Comment (ajax-gnatiuk-s) [reply build.gradle.kts:41]: let's group)
Thanks @OlehSarapuk

File: src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetCsvController.kt [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import com.google.protobuf.ByteString
+import com.google.protobuf.Parser
+import io.nats.client.Connection
+import org.springframework.stereotype.Component
+
+@Component
+class NatsCryptocurrencyGetCsvController(
+    private val cryptocurrencyService: CryptocurrencyService,
+    override val connection: Connection
+) : NatsController<CryptocurrencyRequest, CryptocurrencyResponse> {
```
[the remark this answers is not in the document]
Comment (ajax-gnatiuk-s) [reply src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetCsvController.kt:17]: Nooo, everything is correct here.
```kotlin
class MyClass(r: Runnable) : Runnable {} // CORRECT
class MyClass(r: Runnable): Runnable  {} // INCORRECT


fun myFun(): String = "" // CORRECT
fun myFun() : String = "" // INCORRECT
```

File: src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetAllController.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import com.google.protobuf.Parser
+import io.nats.client.Connection
+import org.springframework.stereotype.Component
+
+@Component
+class NatsCryptocurrencyGetAllController (
+    private val cryptocurrencyService: CryptocurrencyService,
+    private val cryptocurrencyConvertor: CryptocurrencyConvertor,
+    override val connection: Connection
+) : NatsController<CryptocurrencyRequest, CryptocurrencyResponse> {
```
[the remark this answers is not in the document]
Comment (ajax-gnatiuk-s) [reply src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetAllController.kt:18]: everything is correct here :)

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**